### PR TITLE
Redefine scheduler to set learning rate using recursive formula

### PR DESCRIPTION
**Pull request by chandlerzuo**
Modified step_lr for StepLR, MultiStepLR, ExponentialLR and CosineAnnealingLR. In this way, multiple schedulers can be used simultaneously to modify the learning rates. 

Related issue: https://github.com/pytorch/pytorch/issues/13022 

Added unit tests combining multiple schedulers.

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -154,8 +154,12 @@ def __init__(self, optimizer, step_size, gamma=0.1, last_epoch=-1):
...
 
     def get_lr(self):
-        return [base_lr * self.gamma ** (self.last_epoch // self.step_size)
-                for base_lr in self.base_lrs]
+        if (self.last_epoch == 0) or (self.last_epoch % self.step_size != 0):
+            return [group['lr'] for group in self.optimizer.param_groups]
+        return [group['lr'] * self.gamma
+                for group in self.optimizer.param_groups]
+        # return [base_lr * self.gamma ** (self.last_epoch // self.step_size)
+        #         for base_lr in self.base_lrs]
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: nit: I can see why you want to leave the old version of the code commented out (the non-base version is harder to understad, because you're defining the transitions, rather than the end state)--it would be even better if you explicitly added a comment saying so (otherwise, the reader will be puzzled about what this mysterious commented out code is. Did someone forget to delete? Is it an alternate implementation?)

File: test/test_optim.py [hunk 2]
```diff
@@ -512,6 +596,141 @@ def test_reduce_lr_on_plateau8(self):
                                       threshold=0.1, patience=5, cooldown=5)
         self._test_reduce_lr_on_plateau(scheduler, targets, metrics, epochs)
 
+    def test_compound_step_and_multistep_lr(self):
```
Comment (ezyang) [inline test/test_optim.py]: Out of curiosity, how did you pick which combinations to test? :) (It might also be worth just explicitly testing the old implementations, with no combinations, against the new ones; the new implementations seem more complicated, so it seems helpful to actually test explicitly against the reference implementation, which is simpler.)

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -154,8 +154,12 @@ def __init__(self, optimizer, step_size, gamma=0.1, last_epoch=-1):
         super(StepLR, self).__init__(optimizer, last_epoch)
 
     def get_lr(self):
-        return [base_lr * self.gamma ** (self.last_epoch // self.step_size)
-                for base_lr in self.base_lrs]
+        if (self.last_epoch == 0) or (self.last_epoch % self.step_size != 0):
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: The special case for `last_epoch == 0` is a real stick up my craw. The problem here is an existing infelicity in LR design, also noted by a user in https://github.com/pytorch/pytorch/issues/8837

Unfortunately, fixing this would be an even bigger BC-breaking change. Nothing you have to do for this PR, but something good to keep in mind.

CC @soumith

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -183,16 +187,21 @@ class MultiStepLR(_LRScheduler):
     """
 
     def __init__(self, optimizer, milestones, gamma=0.1, last_epoch=-1):
-        if not list(milestones) == sorted(milestones):
-            raise ValueError('Milestones should be a list of'
-                             ' increasing integers. Got {}', milestones)
+        milestones = set(milestones)
+        #if not list(milestones) == sorted(milestones):
+        #    raise ValueError('Milestones should be a list of'
+        #                     ' increasing integers. Got {}', milestones)
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: What happens if a milestone shows up twice in the list?

**Comment by ezyang** [comment]
Changes LGTM.

**Comment by ezyang** [comment]
Nit: When naming a PR, name it for what it does, not a URL (that you have to click through to find out what it's about ;)

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (chandlerzuo) [reply torch/optim/lr_scheduler.py]: I am sorry I actually forgot to delete these lines. I think a better way is to delete the old implementation altogether, while highlighting the change in logic in the description of each function. Do you agree?

File: test/test_optim.py (hunk 2 above)
Comment (chandlerzuo) [reply test/test_optim.py]: I added 10 tests. First, I tested all combinations between StepLR, MultiStepLR, ExponentialLR and CosineAnnealing (6 combinations). Then I tested ReduceLROnPlateau + 1 of these modified changes (4 in total). The purpose of these tests is to see that combining different schedulers now works as intended.

For single implementations, there are already uni-tests for old implementations, and the new implementations already pass those tests, so that gives some validation. But, I can add some tests that implements the old codes explicitly for comparison as you suggested.

File: torch/optim/lr_scheduler.py (hunk 4 above)
Comment (chandlerzuo) [reply torch/optim/lr_scheduler.py]: Good question. I actually think raising an error makes more sense, but the old codes decrease the rate twice at that milestone. Shall we make it BC?

Comment (ezyang) [reply torch/optim/lr_scheduler.py]: It doesn't seem... too hard to maintain BC, so might as well.

**Review by ezyang — APPROVED** [review]

File: torch/optim/lr_scheduler.py (hunk 4 above)
Comment (chandlerzuo) [reply torch/optim/lr_scheduler.py]: I already made the updated version to maintain BC.

**Comment by jeanm** [comment]
Is this blocked on something? Would be great to have it merged!
